pkg/agent/controller/activeprobe/activeprobe_controller.go

```diff
@@ -0,0 +1,271 @@
+		DstMac:     dstMac,
+		IPProtocol: uint8(ap.Spec.Packet.IPHeader.Protocol),
+		IPLength:   ap.Spec.Packet.Length,
+		IPFlags:    uint16(ap.Spec.Packet.IPHeader.Flags),
+		TTL:        uint8(ap.Spec.Packet.IPHeader.TTL),
+	}
+
+	switch packet.IPProtocol {
+	case protocol.Type_ICMP:
+		packet.ICMPEchoID = uint16(ap.Spec.Packet.TransportHeader.ICMP.ID)
+		packet.ICMPEchoSeq = uint16(ap.Spec.Packet.TransportHeader.ICMP.Sequence)
+	case protocol.Type_TCP:
+		packet.SrcPort = uint16(ap.Spec.Packet.TransportHeader.TCP.SrcPort)
+		packet.DstPort = uint16(ap.Spec.Packet.TransportHeader.TCP.DstPort)
+		packet.TCPFlags = uint8(ap.Spec.Packet.TransportHeader.TCP.Flags)
+	case protocol.Type_UDP:
+		packet.SrcPort = uint16(ap.Spec.Packet.TransportHeader.UDP.SrcPort)
+		packet.DstPort = uint16(ap.Spec.Packet.TransportHeader.UDP.DstPort)
+	}
+
+	return packet
+}
+
+func (a *Controller) SendActiveProbePacket(ap *activeprobev1alph1.ActiveProbe) error {
+	var err error
+	ovsbrName := ap.Spec.Source.BridgeName
+	inport := uint32(ap.Spec.Source.Ofport)
+	tag := ap.Status.Tag
+	packet := a.ParseActiveProbeSpec(ap)
+	sendTimes := ap.Spec.ProbeTimes
+
+	for i := 0; i < int(sendTimes); i++ {
+		err = a.DatapathManager.SendActiveProbePacket(ovsbrName, *packet, tag, inport, nil)
+	}
+
+	return err
+}
+
+func (a *Controller) InstallActiveProbeRuleFlow(ovsbrName string, tag uint8, ipDa *net.IP) error {
+	var err error
+	err = a.DatapathManager.InstallActiveProbeFlows(ovsbrName, tag, ipDa)
+	return err
+}
+
+func (a *Controller) updateActiveProbeStatus(ap *activeprobev1alph1.ActiveProbe, apResult *activeprobev1alph1.AgentProbeResult, reason string) error {
+	update := ap.DeepCopy()
+	if reason != "" {
+		update.Status.Reason = reason
+	}
+	update.Status.Results = append(update.Status.Results, *apResult)
+	err := a.K8sClient.Status().Update(context.TODO(), update, &client.UpdateOptions{})
+	klog.Infof("Updated ActiveProbe %s: %+v", ap.Name, update.Status)
+	fmt.Printf("Updated ActiveProbe %s: %+v\n", ap.Name, update.Status)
+	return err
+}
+
+func (a *Controller) deleteActiveProbeByName(apName string) *activeProbeState {
+	a.RunningActiveprobeMutex.Lock()
+	defer a.RunningActiveprobeMutex.Unlock()
+	for tag, apState := range a.RunningActiveprobe {
+		if apState.name == apName {
+			delete(a.RunningActiveprobe, tag)
+			return apState
+		}
+	}
+	return nil
+}
+
+func (a *Controller) cleanupActiveProbe(ap *activeprobev1alph1.ActiveProbe) {
+	var ovsbrName string
+
+	curAgentName := utils.CurrentAgentName()
+	if curAgentName == ap.Spec.Source.AgentName {
+		ovsbrName = ap.Spec.Source.BridgeName
+	} else if curAgentName == ap.Spec.Destination.AgentName {
+		ovsbrName = ap.Spec.Destination.BridgeName
+	}
+
+	apState := a.deleteActiveProbeByName(ap.Name)
+
+	if apState != nil {
+		err := a.DatapathManager.UninstallActiveProbeFlows(ovsbrName, apState.tag)
+		if err != nil {
+			klog.Errorf("Failed to uninstall Traceflow %s flows: %v", apState.name, err)
+		}
+	}
```

Comment:
Traceflow? This logic can combine with deleteActiveProbeByname

Reply:
has been modified